src/alloc/isolated_alloc.rs

```diff
@@ -0,0 +1,356 @@
+            let size_pinfo = size / COMPRESSION_FACTOR;
+            // DenseBitSet::contains() panics if the index is out of bounds
+            if pinfo.domain_size() < idx_pinfo + size_pinfo {
+                break;
+            }
+            // FIXME: is there a more efficient way to check whether the entire range is unset
+            // in the bitset?
+            let range_avail = !(idx_pinfo..idx_pinfo + size_pinfo).any(|idx| pinfo.contains(idx));
+            if range_avail {
+                pinfo.insert_range(idx_pinfo..idx_pinfo + size_pinfo);
+                unsafe {
+                    let ptr = page.add(idx);
+                    if zeroed {
+                        // Only write the bytes we were specifically asked to
+                        // zero out, even if we allocated more
+                        ptr.write_bytes(0, layout.size());
+                    }
+                    return Some(ptr);
+                }
+            }
+        }
+        None
+    }
+
+    /// Allocates in multiples of one page on the host system.
+    ///
+    /// SAFETY: Same as `alloc()`.
+    unsafe fn alloc_huge(&mut self, layout: Layout, zeroed: bool) -> *mut u8 {
+        let layout = IsolatedAlloc::huge_normalized_layout(layout, self.page_size);
+        let ret =
+            unsafe { if zeroed { alloc::alloc_zeroed(layout) } else { alloc::alloc(layout) } };
+        self.huge_ptrs.push((ret, layout.size()));
+        ret
+    }
+
+    /// Deallocates a pointer from this allocator.
+    ///
+    /// SAFETY: This pointer must have been allocated by calling `alloc()` (or
+    /// `alloc_zeroed()`) with the same layout as the one passed on this same
+    /// `IsolatedAlloc`.
+    pub unsafe fn dealloc(&mut self, ptr: *mut u8, layout: Layout) {
+        let (size, align) = IsolatedAlloc::normalized_layout(layout);
+
+        if IsolatedAlloc::is_huge_alloc(size, align, self.page_size) {
+            unsafe {
+                self.dealloc_huge(ptr, layout);
+            }
+        } else {
+            // Offset of the pointer in the current page
+            let ptr_idx = ptr.addr() % self.page_size;
+            // And then the page's base address
+            let page_addr = ptr.addr() - ptr_idx;
+
+            // Find the page this allocation belongs to.
+            // This could be made faster if the list was sorted -- the allocator isn't fully optimized at the moment.
+            let pinfo = std::iter::zip(&mut self.page_ptrs, &mut self.page_infos)
+                .find(|(page, _)| page.addr() == page_addr);
+            let Some((_, pinfo)) = pinfo else {
+                panic!(
+                    "Freeing in an unallocated page: {ptr:?}\nHolding pages {:?}",
+                    self.page_ptrs
+                )
+            };
+            // Since each bit of the bitset represents COMPRESSION_FACTOR bytes,
+            // we need to divide by that when getting the indices
```

Comment:
The most important part is to explain what it is you are doing, not how you are doing it.
```suggestion
            // Mark this range as available in the page.
```